src/System.Collections/src/System/Collections/Generic/HashSet.cs

```diff
@@ -123,7 +123,27 @@ public HashSet(IEnumerable<T> collection, IEqualityComparer<T> comparer)
             }
             Initialize(suggestedCapacity);
 
-            this.UnionWith(collection);
+            var otherAsHashSet = collection as HashSet<T>;
+            if (otherAsHashSet != null && AreEqualityComparersEqual(this, otherAsHashSet))
+            {
+                _count = otherAsHashSet._count;
+                _lastIndex = otherAsHashSet._lastIndex;
+                _freeList = otherAsHashSet._freeList;
+                _version = otherAsHashSet._version;
+
+                _buckets = new int[otherAsHashSet._buckets.Length];
+                otherAsHashSet._buckets.CopyTo(_buckets, 0);
+
+                _slots = new Slot[otherAsHashSet._slots.Length];
+                otherAsHashSet._slots.CopyTo(_slots, 0);
```

Comment:
Yeah, I just tried it out, I suggest we go with a small tweak like I mentioned, e.g.
```
var otherAsHashSet = collection as HashSet<T>;
if (otherAsHashSet != null && AreEqualityComparersEqual(this, otherAsHashSet))
{
    _count = otherAsHashSet._count;
    _lastIndex = otherAsHashSet._lastIndex;
    _freeList = otherAsHashSet._freeList;
    _version = otherAsHashSet._version;
    _buckets = (int[])otherAsHashSet._buckets.Clone();
    _slots = (Slot[])otherAsHashSet._slots.Clone();

    // _comparer is already the same
}
else
{
    // to avoid excess resizes, first set size based on collection's count. Collection
    // may contain duplicates, so call TrimExcess if resulting hashset is larger than
    // threshold
    int suggestedCapacity = 0;
    ICollection<T> coll = collection as ICollection<T>;
    if (coll != null)
    {
        suggestedCapacity = coll.Count;
    }
    Initialize(suggestedCapacity);

    this.UnionWith(collection);
}
```